### ISSUE TITLE
Loss is NaN in training

### ISSUE THREAD

**Issue by DanieleBertagnoli** [aau-cns/poet#21]
Hi, I was trying to train PoET on my custom dataset. I converted it using your script `ycbv2poet.py` since my custom dataset is basically built to emulate YCB-V but with other objects (same number of classes). I also trained the backbone and it works correctly. However, when I try to run the training for PoET (using gt bboxes), it stops immediately for the loss being nan.  I tried to understand which was the problem, but I not sure that these two that I list could be the one:

1. Before stopping, the script prints the following warning (I used the docker image so I didn't touch the dependencies):
```
/opt/conda/lib/python3.8/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  /opt/pytorch/pytorch/aten/src/ATen/native/TensorShape.cpp:2156.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
/opt/project/models/position_encoding.py:53: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  dim_t = self.temperature ** (2 * (dim_t // 2) / self.num_pos_feats)
```

2. I noticed that for some images the returned `targets` in  `PoseDataset` class (in __getitem__) are empty:
```
{'boxes': tensor([], size=(0, 4)), 'labels': tensor([], dtype=torch.int64), 'image_id': tensor([202]), 'relative_position': tensor([], size=(0, 3)), 'relative_quaternions': tensor([], size=(0, 4)), 'relative_rotation': tensor([], size=(0, 3, 3)), 'intrinsics': tensor([], size=(0, 9)), 'area': tensor([]), 'iscrowd': tensor([], dtype=torch.int64), 'orig_size': tensor([480, 640]), 'size': tensor([480, 640])}
```
Actually I think this could be the most probable cause of the loss being nan since I added a print and it always crashes when the target given in input in the forward function are like those. If I double-checked the labels in the original dataset and all these information are present in the json, therefore something appears to be broken in the script `ycbv2poet.py` as these informations are replaces with empty lists.

Best, Daniele

**Comment by tgjantos** [comment]
Hi,

what exactly do you mean by "stops immediately"? Is it training for a couple of images or does it die directly? Point number 2 basically describes the issue. You try to calculate the loss for an input that does not have any target. Are you sure that there exists an annotation for this specific image? It seems that it is not loaded correctly.

Maybe try putting a breakpoint in this [this function of the dataloader](https://github.com/aau-cns/poet/blob/main/data_utils/pose_dataset.py#L117) and step through it. You will probably see what causes the issue.

Best,
Thomas

**Comment by DanieleBertagnoli** [comment]
Thank you for your advice, I spotted a little oversight with my dataset (I was representing boxes as xmin, ymin, xmax, ymax rather than xmin, ymin, width and height). However even if the targets are populated, the loss is still nan.
I noticed that this problem occurs when for a certain frame, the backbone doesn't predict anything, hence predicted bboxes are all [-1,-1,-1,-1] and this seems to be not a valid input for the `matcher.py` that returns an empty list of indices (forward function). 

The final answer should be the following: Since in the script `ycbv2poet.py` we ensure that there is at least one visible object in the frame, the PoET model doesn't admit empty predictions by the backbone. However, since the backbone could fail in detecting something (like in this case, since my object is visible but hard to detect), how should I change the loss computation? 

I confirmed the problem using the gt bboxes, using those the training works as expected.

Best, Daniele

**Comment by tgjantos** [comment]
I guess the straightforward approach would be to discard the image for which no object is predicted. So after the matcher returns an empty list of indices for that image, I would discard the respective GT entries from the target during the loss computation.

Best,
Thomas

**Comment by DanieleBertagnoli** [comment]
Yes actually, this it's what probably I will do. Thank you very much for your supply! Last fast question, in `engine.py` we calculate ADD, ADD-S for each model, however, does ADD-S be computed only for symmetric objects? I my case I have set all symmetries to false in the `symmetries.json` file but ADD-S is still computed for each of those. 

I also noticed that ADD-S computation requires a very long time, while computing ADD takes up to 5 seconds, it is normal?

Best,
Daniele

**Comment by tgjantos** [comment]
There are in total 3 different definitions of the ADD metric:

- ADD: Every object is assumed to be non-symmetric, thus the error can be calculated super fast
- ADD-S: Every object is assumed to be symmetric, hence an minimization of the error has to be performed. This takes quite some time to calculate
- ADD(-S): Depending on the symmetry flag of the object either ADD or ADD-S is computed. 

In your case the ADD and ADD-S metric should return the same values. However, the ADD-S metric will give you better performance as it simply assumes each object to be symmetric. This leads also to the increased computation time that you observe. Hope this helps!

Best,
Thomas
